src/shared/Core/Authentication/OAuth/OAuth2Client.cs

```diff
@@ -184,13 +183,19 @@ public IOAuth2CodeGenerator CodeGenerator
                     $"Missing '{OAuth2Constants.AuthorizationGrantResponse.AuthorizationCodeParameter}' in response.");
             }
 
-            _trace2.WriteRegionLeave((DateTimeOffset.UtcNow - startTime).Seconds,
-                Constants.Trace2Region.Oauth2Category, Constants.Trace2Region.AuthCodeLabel, thread);
-            return new OAuth2AuthorizationCodeResult(authCode, redirectUri, codeVerifier);
+            var result = new OAuth2AuthorizationCodeResult(authCode, redirectUri, codeVerifier);
+            var endTime = (DateTimeOffset.UtcNow - startTime).TotalSeconds;
+            _trace2.WriteRegionLeave(endTime, OAuth2Constants.Trace2Category, label, thread);

```

Comment:
One thing that C# has that I really wish we could do in the C code of Git is the `using` statement. If the trace2 machinery creates an `IDisposable` object that writes the "region enter" event on construction and "region exit" event on disposal, then we can do this kind of pattern:

```cs
using (IDisposable resion = _trace2.StartRegion(...))
{
   // Do stuff
} // exit event written no matter how we exit the block.
```

All of the stuff about thread names, start times, and end times, seem like they would be best encapsulated by that disposable object, to make it incredibly easy for consumers to add regions whenever it seems prudent.

Reply:
I agree this is a much more user-friendly approach. I'll see if I can get it working tomorrow.

Reply:
Interested in your feedback on the implementation.

Reply:
I like the `IDisposable` changes.  I'm wondering if we can make (not necessarily now) a `Region` also implement `ITrace2` and holds a reference to the containing `ITrace2` such that it embeds the former.  Then we could do:

```
foo(_trace2 ITrace2, ...)
{
    using (r1 = _trace2.StartRegion(...))
    {
        using (r2 = r1.StartRegion(...))
        {
            // call a function and pass r2 rather than _trace, for example...
        }
    }
}
```

Going one further, if we also defined that for Threads, where the `using` sets the thread name.  This could be
used in lamba expressions and we wouldn't have to worry about thread-pool vs named vs anonymous threads.

```
foo(_trace2 ITrace2, ...)
{
    using (t1 = _trace2.StartThread("xyzzy", ...))
    {
        using (r1_in_t1 = t1.StartRegion(...))
        {
...
```

This may be too much...